core/win32/callback.c

```diff
@@ -5642,6 +5642,72 @@ intercept_exception(app_state_at_intercept_t *state)
 #endif
             LOG(THREAD, LOG_ASYNCH, 2, "Translated cxt->Xip "PFX" to "PFX"\n",
                 cxt->CXT_XIP, mcontext.pc);
+
+            /*
+             * i#2144 : we check if this is a single step exception
+             * where we diverted the address.
+             */
+            if (pExcptRec->ExceptionCode == EXCEPTION_SINGLE_STEP) {
+                instr_t instr;
+
+                instr_init(dcontext, &instr);
+                decode(dcontext, pExcptRec->ExceptionAddress, &instr);
+                /* Checks that exception address translate on a popf. */
+                if (instr_get_opcode(&instr) == OP_popf || instr_get_opcode(&instr) == OP_iret) {
+                    instr_reset(dcontext, &instr);
+                    decode(dcontext, faulting_pc, &instr);
+                    /* Checks that exception was generated by a nop. */
+                    if (instr_get_opcode(&instr) == OP_nop) {
+                        LOG(THREAD, LOG_ASYNCH, 2,
+                            "Caught generated single step exception at "PFX"\n",
+                            pExcptRec->ExceptionAddress);
+                        /* Will continue after one byte popf or iret. */
+                        if (instr_get_opcode(&instr) == OP_popf) {
+                            dcontext->next_tag = mcontext.pc + POPF_LENGTH;
+                        }
+                        else {
+                            /* We get the return address which was poped into ecx. */
+                            dcontext->next_tag = (app_pc) cxt->CXT_XCX;
+                        }
+                        /* Deletes fragment starting at single step exception
+                         * if it exists.
```

Comment:
FIXME: we have a problem where dispatch will use a shared fragment if it exists (sorry didn't think it through on the last review).  We either need to do a flush here (asynch unlink flush is fine) or add some mechanism where dispatch won't look up a shared fragment on this special exit.  The flush would be the least disruptive to the code as it won't need some new mechanism.  With the flush we would of course not need the looup or delete.